### ISSUE TITLE
Deprecating `evaluate(False)`

### ISSUE THREAD

**Issue by oscarbenjamin** [sympy/sympy#25618]
The `evaluate(False)` context manager beaks everything. Users are constantly using it and then finding that it does not do what they wanted it to do. I think that it should either be deprecated or at least documentation should be very clear about the extremely limited circumstances in which it could reasonably be used.

Some recent issues and discussion: gh-25616 gh-25558 https://github.com/sympy/sympy/pull/25515#issuecomment-1678944787 gh-16776 https://github.com/sympy/sympy/pull/24549#issuecomment-1397452683 https://github.com/sympy/sympy/issues/24077#issuecomment-1254824503 gh-20715

It is clear that a lot of users want to be able to work with unevaluated expressions and SymPy should be improved to support that somehow. However `evaluate(False)` is a terrible way of doing it that cannot ever be made to work properly. Far too much of SymPy's internals depend on automatic evaluation for anything nontrivial to work under the conditions of `evaluate(False)`.

Note that I mean `evaluate(False)` here as distinct from `evaluate=False` e.g.:
```python
In [3]: with evaluate(False):
   ...:     e1 = x + x
   ...: 

In [4]: e2 = Add(x, x, evaluate=False)

In [5]: e1
Out[5]: x + x

In [6]: e2
Out[6]: x + x
```
There are problems with passing `evaluate=False` as in the case of `e2` but in principle it is something that could be made to work (I know others disagree). Using `with evaluate(False)` is okay in this particular case where all that is done is to create a single expression inside the `with` block but will break any nontrivial call to almost any SymPy function: that is what should be discouraged.

It is confusing to users that `evaluate(False)` will break any function like `solve`, `integrate`, `simplify` etc because they do not realise that SymPy has no mechanism for disabling its effect in internal code and all of SymPy's internals depend on `evaluate(True)`. It is impossible to write code for things like `solve` etc in such a way that they could work under the effect of `evaluate(False)`. I don't think that it is possible to communicate clearly to most users what are the SymPy operations that could reasonably be expected to succeed under `evaluate(False)`.

It is also impossible to make it so that `evaluate(False)` actually works in the cases where its use might be reasonable because there is no general mechanism that makes it work: the code to check the global variable needs to be added to *every* Basic subclass and there will always be cases where it is not checked.

I think that firstly we should not suggest that the use of `evaluate(False)` is ever a good idea. The docs should simply discourage its use altogether. We should also not try to fix anything so that `evaluate(False)` can be made to work slightly better in any particular case because it will never work properly and will just be a never ending source of disappointment to users.

Secondly I think that if we are going to support unevaluated expressions in any way at all then it should be by passing `evaluate=False` explicitly. The conclusion at https://github.com/sympy/sympy/pull/25515#issuecomment-1678667420 that `evaluate=False` does not need to be provided by `parse_latex_lark` because users can use `evaluate(False)` is entirely backwards. That PR by the way is going to mean that many users will not want to use the Lark-based LaTeX parser because they *really* do want to have parsing without evaluation (CC @wermos @sylee957). I don't know if that change affects `parse_latex` but it will definitely be viewed as a regression by many SymPy users if `parse_latex` no longer supports parsing without evaluation at the time of the SymPy 1.13 release.

I do think that there should be some better way to support unevaluated expressions. It is absolutely something that users want and also absolutely something that SymPy does not handle well. Some SymPy contributors talk about unevaluated expressions as being "broken" or make other disparaging comments about the idea of using them at all. I think it is important to acknowledge that wanting to use unevaluated expression is a very valid usecase. For example SymPy has loads of code for parsing and printing expressions in different formats: those would be very useful for all sorts of things if you could control it so that the parsing and printing did not mess up the expressions by evaluating/altering everything in between.

However while unevaluated expression are an important use case the specific mechanism of `evaluate(False)` is a terrible idea that can never work properly. It cannot be fixed properly so we should not try to fix it and should focus instead on discouraging its use and on creating better alternatives for what users actually want to do.

**Comment by sylee957** [comment]
I’m anyway trying to suggest something like expression builder pattern, to remedy the broken usage of evaluate=False though.

**Comment by sylee957** [comment]
However, I don’t object to someone making patch to fill out evaluate=evaluate for LaTeX parser.

**Comment by sylee957** [comment]
A checklist if you implement ‘evaluate’ correctly

- Avoid using SymPy functions that does not support evaluate, like sqrt, build directly with Pow
- Avoid using +, -, *, / to build expressions, always use Add, Mul,

**Comment by oscarbenjamin** [comment]
> I generally support your blog article that automatic evaluation is something that is definitely a problem, and I expect things are going to be fixed automatically after fixing that.

These things will not be fixed *automatically*. There will still be a need somehow to specify whether or not evaluation is wanted which means that at some level the user will still need to be able to pass the equivalent of `evaluate=False` (although the exact mechanism might be different or the default might be reversed).

**Comment by wermos** [comment]
> However, I don’t object to someone making patch to fill out evaluate=evaluate for LaTeX parser.

Are you talking about the ANTLR-based parser here?

In any case, I don't particularly have a strong opinion on this either way, but if we _do_ go ahead with deprecating `evaluate=False`, I'd like to point out that the tests for the Lark-based LaTeX parser can be simplified considerably and the transformer class for the Lark LaTeX parser would be able to use arithmetic operators (like `+`, `*`) instead of things like `sympy.Add` and `sympy.Mul`.

In short, if we want to do this, we would have to have a sort of task list for the things that need to be changed/updated, so I'm putting the work that would be required for the Lark LaTeX parser on here.

**Comment by oscarbenjamin** [comment]
> if we _do_ go ahead with deprecating `evaluate=False`

Note as I said in the OP that `with evaluate(False)` is not the same as `evaluate=False`. The suggestion here is to deprecate the former but not the latter.

**Comment by sylee957** [comment]
Anyway, I'm trying to suggest 
https://github.com/sympy/sympy/issues/25620
as an alternative of implementing evaluation control for SymPy functions

**Comment by kompre** [comment]
hello,

I'm a fairly new sympy user and I've stumbled upon the `with evaluate(False)` by accident because it was mentioned in some stackoverflow, but I didn't really find any documentation about it. 

I'm using sympy inside a jupyter notebook and I would like to achieve these results:

1. print expressions with the term in the same order as I write them;
2. substitute values into an expression and show an unevaluated form of such expression (as an intermediary step), before the final evaluation.

In both case I tried a mix of `evaluate=False` and `UnevaluatedExpr` method as suggested by [Advanced Expression Manipulation ](https://docs.sympy.org/latest/tutorials/intro-tutorial/manipulation.html#prevent-expression-evaluation), and got mixed results, particularly regarding the point n.2.

To print an expression with same order as I write them I used the `sympify` method passing the `evaluate=False` argument, and  also set `init_printing(order='none')`: 

- here an example with `evaluate=True` (not the wanted result)

![image](https://github.com/sympy/sympy/assets/11796345/ab4050a2-9749-49f0-9ddd-392c082cf800)

- here an example with `evaluate=False` (what I want to print)

![image](https://github.com/sympy/sympy/assets/11796345/e80ba12d-94eb-4f17-b245-8e50b8f1d422)

I've read tough that this string manipulation should be avoided and explicit symbols expression should be used instead, but in that case I'm not able to achieve the same terms order unless I use `Mul `, `Add`, ecc. losing much of code readability in my opinion. 

![image](https://github.com/sympy/sympy/assets/11796345/5b2772fe-508d-4df9-961c-fb9975213327)

I think that `with evaluate(False)` it looks much nicer and readable:

![image](https://github.com/sympy/sympy/assets/11796345/0e134cfc-ceef-489c-b4c5-8f5ad3045af9)


About the 2nd point (_substitution without evaluation_) I got even more meddling results using the `evaluate=false`/`UnevaluatedExpr` method because with some functions don't play well with `UnevaluatedExpr`. 

Let's take for example the `Max()` function:

- passing the `evaluate=False` will prevent the evaluation

![image](https://github.com/sympy/sympy/assets/11796345/f4d48e60-10af-487d-9364-450e7a57fd3f)

- this does not carry when using substitution:

![image](https://github.com/sympy/sympy/assets/11796345/cdc5fb05-3224-49bd-b79d-4d801602f953)

Using an actual expression  it looks like this:

![image](https://github.com/sympy/sympy/assets/11796345/4bb3b2f8-79e2-439d-8a59-60b9dd1d4ecd)

Then substituting a dictionary of values, the expression get evaluated (I think the `evaluate=False` passed to subs() does nothing).

![image](https://github.com/sympy/sympy/assets/11796345/028c8bdb-178e-4272-b67d-31febbab7ef2)

I try then to wrap the substitution in `UnevaluetedExpr`, but I get a `ValueErrror`, because the argument is not comparable:

![image](https://github.com/sympy/sympy/assets/11796345/da81c88c-99e8-423d-992c-a8d81b7e7fce)
![image](https://github.com/sympy/sympy/assets/11796345/e8226a10-32f9-4835-a139-0238dd4fa305)

If instead of a number I have a symbol in the substitution, then I get a valid expression, albeit it looks wonky because:

- denominator in fractions get expressed as negative power (correct, but confusing)
- terms got switched again

![image](https://github.com/sympy/sympy/assets/11796345/738d6af9-9c36-4922-b59f-53b61667bf21)

also if the symbols are passed in an incorrect order, it fails again (I don't know why):

![image](https://github.com/sympy/sympy/assets/11796345/ed9f703c-92c5-456d-98bc-1fbade4be4e4)

![image](https://github.com/sympy/sympy/assets/11796345/be26070e-0ca1-47a6-85f2-23eeb64dea72)


Instead using `with evaluate(False)` I get a result much more close to my desired output:

![image](https://github.com/sympy/sympy/assets/11796345/624d10b2-10e5-4438-8667-54c16eb5d6ba)
![image](https://github.com/sympy/sympy/assets/11796345/bf017e51-12aa-43a1-89c2-1734986e3432)

In my real application, I am also using quantity with `sympy.physics.units`, that looks a little bit odd because an exponent 1 is present, but otherwise is what I am looking for:

![image](https://github.com/sympy/sympy/assets/11796345/d7ca7580-5ad8-4550-8a96-df4430b02092)

To conclude `with evaluate(False)` has given me, so far, the results I was looking for, in a somewhat concise and readable fashion, while I've failed to obtain the same results with `evaluate=False` and `UnevaluatedExpr`, also getting weird errors depending on the order of the substitution terms. I don't know what `with evaluate(False)` could break though.

**Comment by oscarbenjamin** [comment]
In general something better than `with evaluate(False)` should be provided such as a different expression type where none of the methods simplify implicitly. It is difficult to provide that now because "evaluation" is built in to the basic expression types.

For the type of application that you are demonstrating you would almost certainly want to use the other expression type which also should have things like `div` rather than just negative powers.

> I don't know what `with evaluate(False)` could break though.

It is usually fine to use `evaluate(False)` for basic manipulation but many more complicated SymPy functions can break and in some cases even simple operations might break. It is not really possible to give any precise specification of what would or would not break under the effect of `evaluate(False)`.
